core/src/main/java/cz/o2/proxima/scheme/SchemaDescriptors.java

```diff
@@ -0,0 +1,885 @@
+    }
+
+    @Override
+    public SchemaTypeDescriptor<T> toTypeDescriptor() {
+      return SchemaTypeDescriptor.<T>builder().type(getType()).arrayTypeDescriptor(this).build();
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public boolean equals(Object o) {
+      if (this == o) {
+        return true;
+      }
+      if (o instanceof ArrayTypeDescriptor) {
+        return getType().equals(((ArrayTypeDescriptor<?>) o).getType())
+            && getValueType().equals(((ArrayTypeDescriptor<?>) o).getValueType());
+      }
+      return false;
+    }
+
+    @Override
+    public int hashCode() {
+      return 31 + valueDescriptor.hashCode();
+    }
+  }
+
+  /**
+   * Structure type descriptor allows to have fields with type as another descriptor.
+   *
+   * @param <T> structure type
+   */
+  public static class StructureTypeDescriptor<T> extends GenericTypeDescriptor<T>
+      implements WithDefaultValue<T> {
+
+    @Getter final String name;
+    private final Factory<T> defaultValueFactory;
+    @Getter private List<SchemaDescriptors.FieldTypeDescriptor<?>> fields = new ArrayList<>();
+    private transient Map<String, SchemaDescriptors.FieldTypeDescriptor<?>> fieldsByName;
+
+    public StructureTypeDescriptor(
+        String name,
+        List<SchemaDescriptors.FieldTypeDescriptor<?>> fieldsToAdd,
+        Factory<T> defaultValueFactory) {
+      super(AttributeValueType.STRUCTURE);
+      this.name = name;
+      this.defaultValueFactory = defaultValueFactory;
+      fieldsToAdd.forEach(this::addField);
+    }
+
+    public void addField(SchemaDescriptors.FieldTypeDescriptor<?> field) {
+      Preconditions.checkArgument(
+          !getFieldsAsMap().containsKey(field.getName()), "Duplicate field " + field.getName());
+      this.fields.add(field);
+      this.fieldsByName.put(field.getName(), field);
+    }
+
+    @SuppressWarnings("squid:S1452")
+    public Map<String, SchemaDescriptors.FieldTypeDescriptor<?>> getFieldsAsMap() {
+      if (fieldsByName == null) {
+        fieldsByName =
+            fields
+                .stream()
+                .collect(
+                    Collectors.toMap(
+                        SchemaDescriptors.FieldTypeDescriptor::getName, Function.identity()));
+      }
+      return fieldsByName;
+    }
+
+    public boolean hasField(String name) {
+      return getFieldsAsMap().containsKey(name);
+    }
+
+    @SuppressWarnings("squid:S1452")
+    public SchemaDescriptors.FieldTypeDescriptor<?> getField(String name) {
+      return Optional.ofNullable(getFieldsAsMap().getOrDefault(name, null))
+          .orElseThrow(
+              () ->
+                  new IllegalArgumentException(
+                      "Field " + name + " not found in structure " + getName()));
+    }
+
+    @Nullable
+    @Override
+    public T getDefaultValue() {
+      return defaultValueFactory != null ? defaultValueFactory.apply() : null;
```

Comment:
This should not return `null`. It might be the case here, that `defaultValueFactory` be `null`, but then the default value can be auto-generated from the underlying field descriptors into a `Map`.